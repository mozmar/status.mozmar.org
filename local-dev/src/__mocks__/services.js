const services = [
    {
        display: true,
        group: 'group1',
        hasDetail: false,
        id: 1,
        link: 'test link',
        name: 'test service 1',
        order: 1,
        status: 'healthy'
    },
    {
        display: true,
        hasDetail: true,
        id: 2,
        link: 'test link',
        name: 'test service 2',
        order: 2,
        status: 'healthy'
    },
    {
        display: true,
        group: 'group1',
        hasDetail: false,
        id: 3,
        link: 'test link',
        name: 'test service 3',
        order: 3,
        status: 'warning'
    },
    {
        display: true,
        group: 'group2',
        hasDetail: false,
        id: 4,
        link: 'test link',
        name: 'test service 4',
        order: 10,
        status: 'healthy'
    }
];

export default services;
