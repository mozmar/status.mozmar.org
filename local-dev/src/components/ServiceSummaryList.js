import React, { Component } from 'react';

import ServiceSummaryGroup from './ServiceSummaryGroup';


class ServiceSummaryList extends Component {
    static propTypes = {
        services: React.PropTypes.array.isRequired
    }

    DEFAULT_ORDER_VALUE = 10000;

    constructor() {
        super();

        this.generateGroups = this.generateGroups.bind(this);
        this.getGroupOrder = this.getGroupOrder.bind(this);
    }

    generateGroups() {
        var groups = {
            default: {
                order: this.DEFAULT_ORDER_VALUE + 1,
                services: []
            }
        };

        this.props.services.forEach(service => {
            if (!service.display) {
                return;
            }

            if (service.group) {
                if (Object.keys(groups).indexOf(service.group) >= 0) {
                    groups[service.group].services.push(service);
                } else {
                    groups[service.group] = {
                        services: [service]
                    };
                }
            } else {
                groups.default.services.push(service);
            }
        });

        for (var property in groups) {
            if (groups.hasOwnProperty(property)) {
                if (property === 'default') {
                    // Default group has already calculated order.
                    continue;
                }

                groups[property].order = this.getGroupOrder(groups[property].services);
            }
        };

        return groups;
    }

    /**
     * Group order is defined by the component in the group with the minimum
     * order value.
     *
     * Default is DEFAULT_ORDER_VALUE.
     */
    getGroupOrder(services) {
        var order = this.DEFAULT_ORDER_VALUE;

        services.forEach(service => {
            if (service.order && service.order < order) {
                order = service.order;
            }
        });

        return order;
    }

    render() {
        var groups = this.generateGroups();
        var serviceGroups = [];

        for (var property in groups) {
            if (groups.hasOwnProperty(property)) {
                serviceGroups.push(
                    <ServiceSummaryGroup
                        key={property}
                        name={property}
                        services={groups[property].services}
                        order={groups[property].order} />
                );
            }
        }

        serviceGroups.sort((a, b) => {
            // Reverse ordering
            var a_order = a.props.order;
            var b_order = b.props.order;

            if (a_order < b_order) {
                return -1;
            } else if (a_order > b_order) {
                return 1;
            }

            return 0;
        });

        return (
            <div className="service-groups">
                {serviceGroups}
            </div>
        );
    }
}

export default ServiceSummaryList;
