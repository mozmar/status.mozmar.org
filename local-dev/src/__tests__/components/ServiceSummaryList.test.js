import React from 'react';
import renderer from 'react-test-renderer';

import ServiceSummaryList from '../../components/ServiceSummaryList';

import services from '../../__mocks__/services';

test('render ServiceSummaryList', () => {
    const component = renderer.create(
        <ServiceSummaryList services={services} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
