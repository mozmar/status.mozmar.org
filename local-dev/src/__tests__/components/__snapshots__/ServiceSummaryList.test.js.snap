exports[`test render ServiceSummaryList 1`] = `
<div
  className="service-groups">
  <ul
    className="list-group">
    <li
      className="list-group-item list-group-item-info">
      group1
    </li>
    <li
      className="list-group-item">
      <span
        className="status healthy">
        healthy
      </span>
      test service 1
    </li>
    <li
      className="list-group-item">
      <span
        className="status warning">
        warning
      </span>
      test service 3
    </li>
  </ul>
  <ul
    className="list-group">
    <li
      className="list-group-item list-group-item-info">
      group2
    </li>
    <li
      className="list-group-item">
      <span
        className="status healthy">
        healthy
      </span>
      test service 4
    </li>
  </ul>
  <ul
    className="list-group">
    <li
      className="list-group-item">
      <span
        className="status healthy">
        healthy
      </span>
      <a
        onClick={[Function]}
        style={Object {}}>
        test service 2
      </a>
    </li>
  </ul>
</div>
`;
